eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h

```diff
@@ -25,71 +25,97 @@ class DenseTensorAddressCombiner
-    std::vector<size_t>    _accumulatedSize;
+    class AddressContext {
+    public:
+        AddressContext(const eval::ValueType &type);
+        ~AddressContext();
+        size_type dimSize(uint32_t dim) const { return _type.dimensions()[dim].size; }
+        size_type wholeDimStep(uint32_t dim) const { return _accumulatedSize[dim] * dimSize(dim); }
+        size_t index() const {
+            size_t cellIdx(0);
+            for (uint32_t i(0); i < _address.size(); i++) {
+                cellIdx += _address[i]*_accumulatedSize[i];
+            }
+            return cellIdx;
+        }
+        void update(const Address & addr, const Mapping & mapping) {
+            for (const auto & m : mapping) {
+                _address[m.first] = addr[m.second];
+            }
+        }
+
+        const eval::ValueType &_type;
+        std::vector<size_t>    _accumulatedSize;
+        Address                _address;
+
+    };
+    class CellAddressContext : public AddressContext {
+    public:
+        CellAddressContext(const eval::ValueType &type, CellsRef cells) : AddressContext(type), _cells(cells) { }
+        double cell() const { return cell(index()); }
+        double cell(size_t cellIdx) const { return _cells[cellIdx]; }
+    private:
+        CellsRef               _cells;
+    };
+
+
+    CellAddressContext     _rightAddress;
+    AddressContext         _combinedAddress;
+
     Mapping                _left;
     Mapping                _commonRight;
     Mapping                _right;
-    void update(const Address & addr, const Mapping & mapping) {
-        for (const auto & m : mapping) {
-            _combinedAddress[m.first] = addr[m.second];
-        }
-    }
-    double rightCell(size_t cellIdx) const { return _rightCells[cellIdx]; }
-    size_t rightIndex(const Address &address) const {
-        size_t cellIdx(0);
-        for (uint32_t i(0); i < address.size(); i++) {
-            cellIdx += address[i]*_accumulatedSize[i];
-        }
-        return cellIdx;
-    }
+
 public:
-    DenseTensorAddressCombiner(const eval::ValueType &lhs, const eval::ValueType &rhs, CellsRef rhsCells);
+    DenseTensorAddressCombiner(const eval::ValueType &combined, const eval::ValueType &lhs,
+                               const eval::ValueType &rhs, CellsRef rhsCells);
     ~DenseTensorAddressCombiner();
-    void updateLeftAndCommon(const Address & addr) { update(addr, _left); }
+    void updateLeftAndCommon(const Address & addr) { _combinedAddress.update(addr, _left); }
 
     bool hasAnyRightOnlyDimensions() const { return ! _right.empty(); }
 
-    const Address &address() const { return _combinedAddress; }
+    const Address &address() const { return _combinedAddress._address; }
 
     bool updateCommonRight() {
         for (const auto & m : _commonRight) {
-            if (_combinedAddress[m.first] >= _rightType.dimensions()[m.second].size) {
+            if (_combinedAddress._address[m.first] >= _rightAddress.dimSize(m.second)) {
```

Comment:
could conditionally updating the common parts of the right address be refactored into a function on the AddressContext class? maybe it should even be called from inside the for_each (for_each_matching_right_only?) function?, removing a condition from the outer loop.